auction-server/src/auction/service/simulator.rs

```diff
@@ -0,0 +1,336 @@
+        },
+        account_utils::StateMut,
+        address_lookup_table::state::AddressLookupTable,
+        bpf_loader_upgradeable::UpgradeableLoaderState,
+        commitment_config::CommitmentConfig,
+        pubkey::Pubkey,
+        signature::Signature,
+        slot_hashes::SlotHashes,
+        transaction::VersionedTransaction,
+    },
+    std::{
+        collections::{
+            HashMap,
+            HashSet,
+        },
+        time::Instant,
+    },
+    time::Duration,
+    tokio::sync::RwLock,
+};
+
+pub struct Simulator {
+    receiver:      RpcClient,
+    pending_txs:   RwLock<Vec<(VersionedTransaction, Instant)>>,
+    account_cache: RwLock<HashMap<Pubkey, (Account, Instant)>>,
+}
+
+struct AccountsConfig {
+    accounts:            HashMap<Pubkey, Account>,
+    programs:            HashMap<Pubkey, Account>,
+    upgradable_programs: HashMap<Pubkey, Account>,
+}
+
+impl AccountsConfig {
+    fn new() -> Self {
+        Self {
+            accounts:            Default::default(),
+            programs:            Default::default(),
+            upgradable_programs: Default::default(),
+        }
+    }
+
+    fn apply(&self, svm: &mut LiteSVM) {
+        for (key, account) in self.accounts.iter() {
+            if let Err(e) = svm.set_account(*key, account.clone()) {
+                tracing::error!("Failed to set account for key {:?} {:?}", key, e);
+            }
+        }
+        for (key, account) in self.upgradable_programs.iter() {
+            svm.add_program(
+                *key,
+                &(account.data()[UpgradeableLoaderState::size_of_programdata_metadata()..]),
+            );
+        }
+        for (key, account) in self.programs.iter() {
+            svm.add_program(*key, &account.data);
+        }
+    }
+}
+
+const MAX_PENDING_DURATION: Duration = Duration::seconds(15);
+const ACCOUNT_CACHE_DURATION: Duration = Duration::hours(1);
+
+impl Simulator {
+    pub fn new(receiver: RpcClient) -> Self {
+        Self {
+            receiver,
+            pending_txs: Default::default(),
+            account_cache: Default::default(),
+        }
+    }
+
+    pub async fn fetch_pending_and_remove_old_txs(&self) -> Vec<VersionedTransaction> {
+        let now = Instant::now();
+        let mut pending_txs = self.pending_txs.write().await;
+        pending_txs.retain(|(_, time)| now.duration_since(*time) < MAX_PENDING_DURATION);
+        pending_txs.iter().map(|(tx, _)| tx.clone()).collect()
+    }
+
+
+    pub async fn add_pending_transaction(&self, tx: &VersionedTransaction) {
+        let now = Instant::now();
+        self.pending_txs.write().await.push((tx.clone(), now));
+    }
+
+    pub async fn remove_pending_transaction(&self, sig: &Signature) {
+        self.pending_txs
+            .write()
+            .await
+            .retain(|tx| !tx.0.signatures[0].eq(sig));
+    }
+
+    async fn try_get_accounts_from_cache(&self, keys: &[Pubkey]) -> Option<Vec<Option<Account>>> {
```

Comment:
Why return Vec<Option<Account>>?
You can return Vec<Account> here

Reply:
Maybe you can write a comment for this functon. It'll return if all of accounts exist in cache, else return None